specifyweb/specify/views.py

```diff
@@ -391,3 +392,84 @@ def set_admin_status(request, userid):
+                                "description": "The old AgentID value of the agent record that is to be replaced by the new one."
+                            },
+                            "new_agent_id": {
+                                "type": "integer",
+                                "description": "The new AgentID value of the agent that is replacing the old one."
+                            }
+                        },
+                        'required': ['old_agent_id', 'new_agent_id'],
+                        'additionalProperties': False
+                    }
+                }
+            }
+        },
+        "responses": {
+            "204": {"description": "Success",},
+            "400": {"description": "The AgentID specified does not exist."}
+        }
+    },
+})
+@login_maybe_required
+@require_POST
+def agent_record_replacement(request: http.HttpRequest, old_agent_id, new_agent_id: int) -> http.HttpResponse:
+    """Replaces all the foreign keys referencing the old AgentID
+    with the new AgentID, and deletes the old agent record.
+    """
+    check_permission_targets(request.specify_collection.id, request.specify_user.id, [])
+
+    # Create database connection cursor
+    cursor = connection.cursor()
+
+    with transaction.atomic():
+        # Check to make sure both the old and new agent IDs exist in the table
+        cursor.execute("SELECT * FROM specify.agent WHERE AgentID = %s;", [old_agent_id])
+        if len(cursor.fetchall()) == 0:
+            return http.HttpResponseBadRequest("AgentID: " + old_agent_id + " does not exist.")
+        cursor.execute("SELECT * FROM specify.agent WHERE AgentID = %s;", [new_agent_id])
+        if len(cursor.fetchall()) == 0:
+            return http.HttpResponseBadRequest("AgentID: " + new_agent_id + " does not exist.")
+
+        # Get all of the columns in all of the tables of specify the are foreign keys referencing AgentID
+        sql_get_cols_ref_agent_id = """
+        SELECT TABLE_NAME, COLUMN_NAME
+        FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE
+        WHERE
+        REFERENCED_TABLE_SCHEMA = 'specify' AND
+        REFERENCED_TABLE_NAME = 'agent' AND
+        REFERENCED_COLUMN_NAME = 'AgentID'
+        ORDER BY TABLE_NAME;
+        """
+        cursor.execute(sql_get_cols_ref_agent_id)
+        foreign_key_cols = cursor.fetchall()
+        
+        # Build query to update all of the records with foreign keys referencing the AgentID
+        sql_update = ""
+        for table_name, column_names in groupby(foreign_key_cols, lambda x: x[0]):
+            for col in [c[1] for c in column_names]:
+                sql_set_clause = col + " = " + new_agent_id
+                sql_where_clause = col + " = " + old_agent_id
+                sql_update += "UPDATE " + table_name + " SET " + sql_set_clause + " WHERE " + sql_where_clause + ";\n"
+        
+        # Execute update query for agent children
+        cursor.execute(sql_update)
```

Comment:
Could this fail due to business rule errors? @benanhalt do you know how to catch that?
Should return some sort of 4xx error to the front-end in that case

Reply:
Yes. BRs that are enforced by MySQL could occur here. They will probably appear as django.db.IntegrityError or some such. I would approach this by starting to write some tests to see what can happen. You can take a look at https://github.com/specify/specify7/blob/production/specifyweb/specify/api_tests.py#L514-L532 for an example where I was testing an API using the Django "test client" to simulate requests.

One case where I know you can get such an exception to occur is if a CollectingEvent has two different Collectors that get merged to the same agent. You could write a test to set up such a situation and see what exception is produced and what information can be gleaned from it; hopefully at least which table was involved.